### ISSUE TITLE
Implement an API for the Power Management Controller (PMC)

### ISSUE THREAD

**Issue by tmplt** [atsams-rs/atsamx7x-rust#5]
The PMC is a fundamental component of this MCU family as it controls all system and user peripheral clocks. A solid HAL will require a PMC API which should bookkeep the clock configuration state in order to configure other peripherals.

The clock distribution is not _too_ crazy (§31.3 from the family data sheet):
![2022-03-14T15:54:39+01:00](https://user-images.githubusercontent.com/6118602/158198492-1c4b2258-1f76-4e59-b5a4-b799e5511b8e.png)

**Comment by michalfita** [comment]
Preliminary draft of the approach:
```rust
trait SlowClock {
    const fn get_ckgr_mor() -> bit; // Maybe that could reach right register straight away
}
struct SlowRCOscillator;
impl SlowClock for SlowRCOscillator() {
    const fn get_ckgr_mor() -> bit { 0 }
}
struct SlowCrystalOscillator;
impl SlowClock for SlowCrystalOscillator {
    const fn get_ckgr_mor() -> bit { 1 }
}
trait MainClock {
    // ...
}
struct ClockGenerator<SC: SlowClock, MC: MainClock> {
    // ...
}
```
Actually more I think about it, more direct approach to right registers would let us easily implement conversion from one setup of `ClockGenerator` to another.

Similar approach to other parts as well. We need mechanism to get these from system-wide singletons referring to out-of-the-box settings (after reboot), and then convert in the program to desired variant.

The example is flawed as I recalled looking at the diagram, both Main Clock and Slow Clock are active at the same time and Host Clock Controller switches the feed of MCK. Getting to the point were type safety guarantees right clocks setup for all peripherals look non-trivial.

I have a problem with `const fn` for Rust 1.46 - do we have to stick with that version?

**Comment by tmplt** [comment]
I'll see about taking a look at some other clocking APIs for inspiration. [The WIP API for atsamd](https://github.com/atsamd-rs/atsamd/pull/450) may be of interest.

> I have a problem with const fn for Rust 1.46 - do we have to stick with that version?

I see no reason to. Let's target the stable channel for now.

**Comment by jamesmunns** [comment]
As a note, I have a [rudimentary](https://github.com/jamesmunns/same70-experiments/blob/main/vendor/atsamx7x-hal/src/pmc.rs) implementation of PMC support that might make sense to upstream at some point.

I do plan to upstream the whole repo's contents, but I'm working on it as a fork until I complete the actual objectives of my current contract work.

**Comment by tmplt** [comment]
Closed via #25.
